ui/packages/consul-ui/app/services/repository/peer.js

```diff
@@ -50,14 +50,18 @@ export default class PeerService extends RepositoryService {
       });
   }
 
+  @dataSource('/:partition/:ns/:dc/peer-generate/')
+  @dataSource('/:partition/:ns/:dc/peer-initiate/')
```

Comment:
It looks like we were relying on this data-source for the creation and the edit case. I.e. when `name` is an empty string, we will return a newly created `Peer`-model if there's already a name we will fetch from the API.

Two questions:

1. If we wanted to stick to the data-source to generate a new ember-data model instance, can we get rid of the `name === ''` check and create a new method that only creates a record and annotate this with these new annotations? This would clean up the `fetchOne` method to only request from the backend and probably make it easier to understand what's going on. One method fetches data and the other one creates a new model instance.
2. Wouldn't it be easier to not rely on the data-source here and put the logic of creating a new peer model and cleaning it up if need be into the form itself?

Reply:
> 1. If we wanted to stick to the data-source to generate a new ember-data model instance, can we get rid of the name === '' check and create a new method that only creates a record and annotate this with these new annotations? This would clean up the fetchOne method to only request from the backend and probably make it easier to understand what's going on. One method fetches data and the other one creates a new model instance.

We could do, but personally I don't feel like a single conditional here makes the method harder to understand, not only that but whether you are creating a 'new' record or not, you are still being returned a 'filled out' record, so them both being the same thing makes sense to me, they both 'fetch' new data from the repository. The fact that one makes a HTTP request and one doesn't is hidden by the repository layer.

> 2. Wouldn't it be easier to not rely on the data-source here and put the logic of creating a new peer model and cleaning it up if need be into the form itself?

I would say no it wouldn't be easier. This is how the rest of the app works, we pass data into the components from the outside. We avoid loading any sort of data deep down inside component trees. All our components are largely presentational.

Our abstract/base repository class has this very same behaviour and its used by all our models.

https://github.com/hashicorp/consul/blob/155c4bc2af77f71ca9cd5067f3c18d1c9ff67ecf/ui/packages/consul-ui/app/services/repository.js#L186-L193